resource/file/file.go

```diff
@@ -0,0 +1,566 @@
+
+	// TypeFile is a regular file
+	TypeFile Type = "file"
+
+	// TypeLink is a hardlink
+	TypeLink Type = "hardlink"
+
+	// TypeNone is an undefined type
+	TypeNone Type = ""
+
+	// TypeSymlink is a symlink
+	TypeSymlink Type = "symlink"
+
+	// DefaultType is the default type
+	DefaultType Type = TypeFile
+)
+
+var (
+	// ValidFileTypes are valid file types
+	ValidFileTypes = []Type{TypeDirectory, TypeFile}
+
+	// ValidLinkTypes are valid link types
+	ValidLinkTypes = []Type{TypeLink, TypeSymlink}
+
+	// AllTypes are the current FileTypes that are supported
+	AllTypes = append(ValidFileTypes, ValidLinkTypes...)
+)
+
+func (t Type) String() string {
+	return string(t)
+}
+
+// File contains information for managing files
+type File struct {
+	Destination string
+	State       State
+	Type        Type
+	Target      string
+	Force       bool    //force replacement of file
+	Mode        *uint32 //requested permissions from Prepare
+	UserInfo    *user.User
+	GroupInfo   *user.Group
+	Content     []byte
+
+	action Action //create, delete, modify
+
+	renderer resource.Renderer
+}
+
+// New returns a File with UserInfo and GroupInfo fields allocated
+func New() *File {
+	return &File{
+		UserInfo:  &user.User{},
+		GroupInfo: &user.Group{},
+	}
+}
+
+// Apply changes to file resources
+func (f *File) Apply() (resource.TaskStatus, error) {
+
+	status, err := f.diff()
+	if err != nil || status.Level == resource.StatusFatal {
+		return status, errors.Wrap(err, "apply")
+	}
+
+	switch f.action {
+	case ActionDelete:
+		err = f.Delete()
+	case ActionCreate, ActionModify:
+		err = f.Modify(status)
+	}
+	if err != nil {
+		status.Level = resource.StatusFatal
+		status.AddMessage(err.Error())
+	}
+	return status, err
+}
+
+// Check File settings
+func (f *File) Check(r resource.Renderer) (resource.TaskStatus, error) {
+
+	f.renderer = r
+
+	status, err := f.diff()
+	status.AddMessage(fmt.Sprintf("%s (%s)", f.Destination, f.Type))
+	if err != nil || status.Level == resource.StatusFatal {
+		return status, errors.Wrap(err, "check")
+	}
+
+	return status, err
+}
+
+// Compare desired state with an actual file (if present)
+func (f *File) diff() (*resource.Status, error) {
+
+	status := &resource.Status{}
+	var stat os.FileInfo
+	err := f.Validate()
+
+	if err != nil {
+		return status, err
+	}
+
+	stat, err = os.Lstat(f.Destination)
+
+	if os.IsNotExist(err) { // file not found
+		switch f.State {
+		case StateAbsent: // if "absent" is set and the file doesn't exist, return with no changes
+			return status, nil
+
+		case StatePresent: // file doesn't exist, we need to create it
+			f.action = ActionCreate
+			actual := &File{
+				Destination: "<file does not exist>",
+				State:       StateAbsent,
+				Type:        f.Type,
+				UserInfo:    &user.User{},
+				GroupInfo:   &user.Group{},
+			}
+			err = f.diffFile(actual, status)
+			if err != nil {
+				status.Level = resource.StatusFatal
+				return status, errors.Wrapf(err, "diff")
```

Comment:
this error will be displayed to the user. "diff: errmsg" doesn't strike me as being particularly useful. Also, you can use `errors.Wrap` here since you don't have any formatting directives in the message.